src/dsp_tools/commands/excel2json/new_lists.py

```diff
@@ -12,47 +13,116 @@
+    return _fill_parent_id_col_all_excels(all_file_dict)
+
+
+def _fill_parent_id_col_all_excels(excel_dfs: dict[str, dict[str, pd.DataFrame]]) -> dict[str, dict[str, pd.DataFrame]]:
+    all_file_dict = {}
+    for filename, sheets in excel_dfs.items():
+        single_file_dict = {}
+        for sheet_name, df in sheets.items():
+            single_file_dict[sheet_name] = _fill_parent_id_col_one_df(df, _get_preferred_language(df.columns))
+        all_file_dict[filename] = single_file_dict
+    return all_file_dict
+
+
+def _fill_parent_id_col_one_df(df: pd.DataFrame, preferred_language: str) -> pd.DataFrame:
+    """Create an extra column with the ID of the parent node."""
+    # To start, all rows get the ID of the list. These will be overwritten if the row has another parent.
+    df["parent_id"] = df.at[0, "id"]
+    columns = _get_columns_of_preferred_lang(df.columns, preferred_language, r"\d+")
+    for col in columns:
+        grouped = df.groupby(col)
+        for name, group in grouped:
+            if group.shape[0] > 1:
+                # The first row already has the correct ID assigned
+                rest_index = list(group.index)[1:]
+                df.loc[rest_index, "parent_id"] = group.at[group.index[0], "id"]
+    return df
+
+
+def _resolve_duplicate_ids_all_excels(
+    excel_dfs: dict[str, dict[str, pd.DataFrame]],
+) -> dict[str, dict[str, pd.DataFrame]]:
+    ids = []
+    for sheets in excel_dfs.values():
+        for df in sheets.values():
+            ids.extend(df["id"].tolist())
+    counter = Counter(ids)
+    if duplicate_ids := [item for item, count in counter.items() if count > 1]:
+        return _remove_duplicate_ids_in_all_excel(duplicate_ids, excel_dfs)
+    return excel_dfs
+
+
+def _remove_duplicate_ids_in_all_excel(
+    duplicate_ids: list[str], excel_dfs: dict[str, dict[str, pd.DataFrame]]
+) -> dict[str, dict[str, pd.DataFrame]]:
+    for sheets in excel_dfs.values():
+        for df in sheets.values():
+            preferred_lang = _get_preferred_language(df.columns)
+            for i, row in df.iterrows():
+                if row["id"] in duplicate_ids:
+                    df.at[i, "id"] = _construct_non_duplicate_id_string(df.iloc[int(str(i))], preferred_lang)
+    return excel_dfs
+
+
+def _complete_id_one_df(df: pd.DataFrame, preferred_language: str) -> pd.DataFrame:
     if "ID (optional)" not in df.columns:
         df["ID (optional)"] = pd.NA
-    df = _fill_auto_id_column(df, preferred_language)
+    df = _create_auto_id_one_df(df, preferred_language)
     df["id"] = df["ID (optional)"].fillna(df["auto_id"])
-    df = _fill_parent_id(df, preferred_language)
+    df = _resolve_duplicate_ids_keep_custom_change_auto_id_one_df(df, preferred_language)
     return df
 
 
-def _fill_auto_id_column(df: pd.DataFrame, preferred_language: str) -> pd.DataFrame:
+def _resolve_duplicate_ids_keep_custom_change_auto_id_one_df(df: pd.DataFrame, preferred_language: str) -> pd.DataFrame:
```

Comment:
renaming suggestion: _change_auto_id_if_it_is_duplicate_of_custom_id

You don't need the "one df" here, there is no naming collision with other function names

Reply:
But I like consistency, I don't need to wonder what the scope is...

Reply:
Understandable. On the other hand, you can also see that from the arguments list, here it is 1 single df. You can also go for _change_auto_id_if_it_is_duplicate_of_custom_id_for_one_df, if you prefer.

Reply:
I have a suggestion, we skip all the naming nuances and at the very end, we sit together and go through it...

Reply:
Okay, that's fine 😄